### ISSUE TITLE
LMM Registry can shutdown an LLMWrapper

### ISSUE THREAD

**Issue by gflachs** [gflachs/GreenPrompt_LLM_Registry#16]
# Feature: Shutdown LLM Wrapper

## User Story
As a user, I want the LLM Registry to shut down an LLM Wrapper so that resources can be freed or the Wrapper can be taken offline safely when no longer needed.

## Description
The LLM Registry must implement functionality to shut down an LLM Wrapper. This ensures efficient resource management, scaling down operations, or safely decommissioning Wrappers. The process must handle status transitions and ensure resources are released properly.

---

## Requirements
1. **Shutdown Command:**
   - The LLM Registry connects to the machine running the LLM Wrapper via SSH.
   - It terminates the Python process of the LLM Wrapper.
   - It confirms the process has been terminated.

2. **Status Management:**
   - During the shutdown process, the Wrapper's status must be updated to `"shutting_down"`.
   - Once the shutdown is complete, the status is set to `"offline"`.

3. **Error Handling:**
   - If the shutdown process fails, the Registry logs the failure and retries the operation.
   - If the shutdown fails after 3 attempts, the Registry sets the status to `"not_working"`.

4. **Timeouts:**
   - Define a timeout period for the shutdown process. If the Wrapper does not shut down within this time, mark the attempt as failed

---

## Acceptance Criteria

Feature: Shutdown Process for LLM Wrapper

  Scenario: Successful Shutdown of LLM Wrapper
    Given the Registry is managing the LLM Wrapper
    When the shutdown process is initiated
    Then the LLM Wrapper should be successfully shut down via SSH
    And the status should transition from "Shutting Down" to "Offline"

  Scenario: Handling Errors During Shutdown
    Given the shutdown process has failed
    When the error occurs during the shutdown process
    Then the Registry should log the error
    And the Registry should retry the shutdown process up to 3 times
    When 3 failed attempts have occurred
    Then the status should transition to "not_working"

  Scenario: Handling Timeout During Shutdown
    Given the shutdown process exceeds the timeout period
    When the shutdown is not completed within the allowed time
    Then the attempt should be marked as failed

  Scenario: Logging Shutdown Process
    Given the shutdown process has started
    When the shutdown process is completed (success or failure)
    Then all steps, including start, success, failure, and retries, should be logged for auditing purposes

---

### Prerequisites
- Ensure the machine running the LLM Wrapper is accessible via SSH.
- Deploy a test Python process representing the LLM Wrapper.

### Test Cases
1. **Test Successful Shutdown:**
   - Simulate an operational Wrapper.
   - Trigger the shutdown process.
   - Verify the Python process is terminated successfully via SSH commands.
   - Confirm the Wrapper status transitions to `"Offline"` after shutdown.

2. **Test Failed Shutdown:**
   - Simulate a failure scenario where the shutdown process (e.g., SSH or process termination) fails.
   - Verify the Registry retries the operation up to 3 times.
   - Confirm the Wrapper status transitions to `"not_working"` after 3 failed attempts.

3. **Test Timeout Handling:**
   - Simulate a scenario where the Wrapper does not terminate within the timeout period.
   - Verify the Registry  marks the shutdown attempt as failure

4. **Test Logging:**
   - Check logs to ensure all shutdown attempts, failures, and status transitions are properly recorded.

---

### Key Concepts to Understand:
1. **SSH Commands:**
   - Use Python libraries like `paramiko` or `subprocess` to send remote commands for process termination.
   - Example SSH command:
     ```bash
     ssh user@host 'pkill -f llm_wrapper.py'
     ```
2. **Process Verification:**
   - Learn to verify whether a process is running using commands like `pgrep` or checking logs.

3. **Retries and Timeouts:**
   - Implement retry logic with a loop and use Python’s `time` module to manage timeouts.

4. **Status Updates:**
   - Ensure the status transitions are logged and handled systematically.

### Simplifications:
- Provide basic command templates:
  ```python
  ssh_command = "ssh user@host 'pkill -f llm_wrapper.py'"
